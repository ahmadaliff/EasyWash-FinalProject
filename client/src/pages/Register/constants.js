export const SET_STEP = 'SET_STEP';
export const SET_EMAIL = 'SET_EMAIL';
export const RESET_STEP = 'RESET_STEP';
export const SET_EXPIRE_TIME = 'SET_EXPIRE_TIME';
