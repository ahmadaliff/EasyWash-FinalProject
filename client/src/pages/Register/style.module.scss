@import '@styles/base';
.main {
  @include directionX;
  position: relative;
  min-height: 100vh;
  align-items: center;
  justify-content: space-around;
  @include apply-at-sm-size {
    justify-content: center;
    box-sizing: border-box;
  }
  .registerIlustration {
    display: none;
    max-width: 100vw;
    padding: 2rem;
    margin: 0 2rem 0 2rem;
    @include apply-at-sm-size {
      display: block;
      height: 40vh;
    }
    @include apply-at-md-size {
      height: 60vh;
    }
    @include apply-at-lg-size {
      height: 90vh;
    }
  }
  .logoHeader {
    position: absolute;
    cursor: pointer;
    width: auto;
    top: 0;
    padding: 1rem;
    height: 2.5rem;
    .logo {
      height: 2.5rem;
    }
  }
  .content {
    @include directionY;
    align-items: center;
    width: 100%;
    @include apply-at-sm-size {
      z-index: 2;
    }
    .registerCard {
      padding-top: 5rem;
      width: 100%;
      height: 100vh;
      @include apply-at-sm-size {
        padding-top: 0;
        height: 100%;
        max-width: 17rem;
      }
      .headerRegister {
        @include directionX;
        align-items: end;
        justify-content: space-between;
        padding: 1.5rem 0 1.5rem 0;
        h3 {
          margin: 0;
        }
        a {
          font-size: small;
          color: var(--color-primary);
          text-decoration: none;
        }
      }
    }
  }
}
