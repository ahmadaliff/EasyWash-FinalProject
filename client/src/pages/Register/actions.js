import { RESET_STEP, SET_EMAIL, SET_EXPIRE_TIME, SET_STEP } from './constants';

export const actionSetStep = (step) => ({
  type: SET_STEP,
  step,
});
export const actionSetEmail = (email) => ({
  type: SET_EMAIL,
  email,
});
export const actionSetExpire = (expire) => ({
  type: SET_EXPIRE_TIME,
  expire,
});
export const actionHandleResetStep = () => ({
  type: RESET_STEP,
});
