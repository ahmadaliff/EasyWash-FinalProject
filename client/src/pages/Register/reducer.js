import { produce } from 'immer';

import { RESET_STEP, SET_EMAIL, SET_EXPIRE_TIME, SET_MERCHANT, SET_ROLE, SET_STEP } from '@pages/Register/constants';

export const initialState = {
  step: 0,
  email: null,
  expire: null,
  role: null,
  merchant: null,
};

export const storedKey = ['step', 'email', 'expire', 'role', 'merchant'];

const registerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_STEP:
        draft.step = action.step;
        break;
      case SET_EMAIL:
        draft.email = action.email;
        break;
      case SET_ROLE:
        draft.role = action.role;
        break;
      case SET_MERCHANT:
        draft.merchant = action.merchant;
        break;
      case SET_EXPIRE_TIME:
        draft.expire = action.expire;
        break;
      case RESET_STEP:
        return initialState;
    }
  });

export default registerReducer;
