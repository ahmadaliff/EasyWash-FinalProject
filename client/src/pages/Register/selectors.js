import { createSelector } from 'reselect';

import { initialState as registerInitialState } from '@pages/Register/reducer';

const selectRegisterState = (state) => state.registerStep || registerInitialState;
export const selectStep = createSelector(selectRegisterState, (state) => state.step);
export const selectEmail = createSelector(selectRegisterState, (state) => state.email);
export const selectExpire = createSelector(selectRegisterState, (state) => state.expire);
