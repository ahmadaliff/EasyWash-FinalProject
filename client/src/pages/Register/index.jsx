import PropTypes from 'prop-types';
import toast from 'react-hot-toast';
import { useEffect } from 'react';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Link, useNavigate } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';

import { Card, CardContent } from '@mui/material';

import VerifyEmail from '@components/VerifyEmail';
import RegisterForm from '@components/RegisterForm';
import RegisterRole from '@components/RegisterRole';
import VerifyEmailOTP from '@components/VerifyEmailOTP';

import { selectLogin } from '@containers/Client/selectors';
import { selectStep } from '@pages/Register/selectors';
import { actionResetRegisterValue } from '@containers/Client/actions';
import { actionHandleResetStep } from '@pages/Register/actions';

import classes from '@pages/Register/style.module.scss';

const Register = ({ isLogin, step, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (isLogin) {
      toast.error(formatMessage({ id: 'app_user_already_login' }));
      navigate('/');
    }
  }, [formatMessage, isLogin, navigate]);

  const renderStep = () => {
    switch (step) {
      case 0:
        return <VerifyEmail />;
      case 1:
        return <VerifyEmailOTP />;
      case 2:
        return <RegisterRole />;
      case 3:
        return <RegisterForm />;
    }
  };

  return (
    <main className={classes.main} data-testid="register-page">
      <div onClick={() => navigate('/')} className={classes.logoHeader} data-testid="logo">
        <img src="longLogo.svg" alt="logo" className={classes.logo} />
      </div>
      <div className={classes.ilustrationWrap}>
        <img src="register.svg" className={classes.registerIlustration} alt="https://storyset.com" />
        <a href="https://storyset.com/web" className={classes.storySet}>
          Web illustrations by Storyset
        </a>
      </div>
      <div className={classes.content}>
        <Card className={classes.registerCard}>
          <CardContent>
            <div className={classes.headerRegister}>
              <h3>
                <FormattedMessage id="app_header_register" />
              </h3>
              <Link
                to="/Login"
                onClick={() => {
                  dispatch(actionResetRegisterValue());
                  dispatch(actionHandleResetStep());
                }}
              >
                <FormattedMessage id="app_header_login" />
              </Link>
            </div>
            {renderStep()}
          </CardContent>
        </Card>
      </div>
    </main>
  );
};

Register.propTypes = {
  intl: PropTypes.object,
  isLogin: PropTypes.bool,
  step: PropTypes.number,
};

const mapStateToProps = createStructuredSelector({
  isLogin: selectLogin,
  step: selectStep,
});

export default injectIntl(connect(mapStateToProps)(Register));
