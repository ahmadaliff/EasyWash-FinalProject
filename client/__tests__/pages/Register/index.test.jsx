import Register from '@pages/Register';
import { render, fireEvent, act } from '@utils/testHelper';

let wrapper;
let mockStep = 0;
let mockLogin = false;
const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
  Link: () => <b />,
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));
jest.mock('@containers/Client/selectors', () => ({
  selectLogin: () => mockLogin,
}));
jest.mock('@pages/Register/selectors', () => ({
  selectStep: () => mockStep,
}));

jest.mock('@components/MapLeaflet', () => jest.fn().mockImplementation(() => <div />));
jest.mock('@components/VerifyEmail', () => jest.fn().mockImplementation(() => <div data-testid="verify-comp" />));
jest.mock('@components/VerifyEmailOTP', () =>
  jest.fn().mockImplementation(() => <div data-testid="verify-otp-comp" />)
);
jest.mock('@components/RegisterRole', () =>
  jest.fn().mockImplementation(() => <div data-testid="register-role-comp" />)
);
jest.mock('@components/RegisterForm', () =>
  jest.fn().mockImplementation(() => <div data-testid="register-form-comp" />)
);

describe('Register Page', () => {
  test('render', () => {
    const { getByTestId } = render(<Register />);
    expect(getByTestId('register-page')).toBeInTheDocument();
  });
  test('navigate home when logo click', async () => {
    const { getByTestId } = render(<Register />);
    const button = getByTestId('logo');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
  describe('step 0', () => {
    test('should render comp verify email', async () => {
      const { getByTestId } = render(<Register />);
      expect(getByTestId('verify-comp')).toBeInTheDocument();
    });
  });
  describe('step 1', () => {
    test('should render comp verify email otp', async () => {
      mockStep = 1;
      const { getByTestId } = render(<Register />);
      expect(getByTestId('verify-otp-comp')).toBeInTheDocument();
    });
  });
  describe('step 2', () => {
    test('should render comp register role', async () => {
      mockStep = 2;
      const { getByTestId } = render(<Register />);
      expect(getByTestId('register-role-comp')).toBeInTheDocument();
    });
  });
  describe('step 3', () => {
    test('should render comp register form', async () => {
      mockStep = 3;
      const { getByTestId } = render(<Register />);
      expect(getByTestId('register-form-comp')).toBeInTheDocument();
    });
  });
  test('should navigate home cause already login', () => {
    mockLogin = true;
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
