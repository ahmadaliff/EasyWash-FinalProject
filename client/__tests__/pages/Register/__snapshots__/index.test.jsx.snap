// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page Matches snapshot 1`] = `undefined`;
